text/0000-thread-safe-handles.md

```diff
@@ -0,0 +1,532 @@
+
+_`Root` references may only be used for `Object` and `Function` types. This is a limitation of `n-api`._
+
+```rust
+/// `Send` Root handle to JavaScript objects
+impl Root<T: Object> {
+    pub fn new<'a, C: Context<'a>>(
+        cx: &mut C,
+        v: &T,
+    ) -> Self;
+
+    pub fn into_inner<'a, C: Context<'a>>(
+        self,
+        cx: &C,
+    ) -> T;
+
+    pub fn drop<'a, C: Context<'a>>(
+        self,
+        cx: &C,
+    );
+}
+```
+
+#### `trait Object`
+
+The idiomatic way to create a `Root<_>` is from an `Object`.
+
+```rust
+trait Object: Value {
+    fn root<'a, C: Context<'a>>(&self, cx: &mut cx) -> Root<Self>;
+}
+```
+
+#### `trait Finalize`
+
+```rust
+impl<T> Finalize for Root<T> {
+    fn finalize<'a, Context<'a>>(self, cx: &mut C) {
+        self.drop(cx);
+    }
+}
+```
+
+`Root<_>` implement `Finalize`to allow transparent dropping within containers; e.g., `Arc`. The following boilerplate can be used to properly dispose of an `Arc<Root<_>>`.
+
+```rust
+struct MyCallback {
+    queue: EventQueue,
+    callback: Arc<Root<JsFunction>>,
+}
+
+impl Finalize for MyStruct {
+    fn finalize<'a, Context<'a>>(self, cx: &mut C) {
+        self.callback.finalize(cx);
+    }
+}
+
+impl MyCallback {
+    fn call<'a, C: Context<'a>>(&self) -> JsResult<JsUndefined> {
+        let queue = self.queue.clone();
+        let callback = Arc::clone(self.callback);
+
+        std::thread::spawn(move || queue.send(move |cx| {
+            let callback = Arc::try_unwrap(callback)
+                .or_else(|cb| cb.clone(&mut cx))
+                .into_inner(&mut cx);
+
+            let this = cx.undefined();
+            let args = vec![cx.string("Hello, World!")];
+            let _ = cb.call(&mut cx, this, args)?;
+
+            Ok(cx.undefined())
+        }));
+    }
+}
+```
+
+#### Design Notes
+
+A `Root<_>` can only be dropped on the JavaScript thread that created it because they include a non-atomic reference count. There are two ways to safely drop a `Root`: `Root::into_inner` and `Root::drop`.
+
+The `Drop` implementation of `Root` will `panic` to alert users of the leak and guide them towards correct usage.
+
+```rust
+impl Drop for Root {
+    fn drop(&mut self) {
+        if !std::thread::panicking() {
+            panic!("neon::handle::Root leaked. Must call into_inner or drop.");
+        }
+    }
+}
+```
+
+In the future, neon may include a higher level construct that abstracts away this error prone usage.
+
+### `neon::unbounded::EventQueue`
+
+```rust
+impl EventQueue {
+    /// Creates an unbounded queue
+    pub fn new<'a, C: Context<'a>>(cx: &mut C) -> Self;
+
+    /// Allow the Node event loop to exit while this `EventQueue` exists.
+    /// _Idempotent_
+    pub fn unref(&mut self);
+
+    /// Prevent the Node event loop from existing while this `EventQueue` exists. (Default)
+    /// _Idempotent_
+    pub fn reference(&mut self);
```

Comment:
@dherman What do you think of returning `-> &mut self` here to allow for chaining? It doesn't have too much value at the moment, but if we add more mutations it could be useful later without requiring a breaking change.